### PR TITLE
[4.0] Use created date in mod popular if publish_up is nulldate

### PR DESCRIPTION
**Pull request by chmst**
### Summary of Changes
Use created date in mod_popular and mod_latest if publish_up is nulldate.

### Testing Instructions
Add a new article, then have a look on molule popular articles and mod_latest on the administrator panel. 

### Expected result
Dates are correct

### Actual result
popular-publish_up

### CODE REVIEW

**Comment by brianteeman** [comment]
The same is true for Recently Added Articles

File: administrator/modules/mod_popular/tmpl/default.php [hunk 1]
```diff
@@ -51,8 +53,7 @@
 						<span class="small">
 							<span class="icon-calendar" aria-hidden="true"></span>
 							<?php 
-								$nulldate = Factory::getDbo()->getNullDate();
-								$date = $item->publish_up == $nulldate ? $item->created : $item->publish_up; 
+								$date = $item->publish_up === $nulldate() ? $item->created : $item->publish_up;
```
Comment (alikon) [inline administrator/modules/mod_popular/tmpl/default.php]: `$nulldate()`   suspect a typo

Comment (chmst) [reply administrator/modules/mod_popular/tmpl/default.php]: oops

File: administrator/modules/mod_latest/tmpl/default.php [hunk 2]
```diff
@@ -46,7 +49,10 @@
 				<span class="badge badge-secondary badge-pill">
 					<span class="small">
 						<span class="icon-calendar" aria-hidden="true"></span>
-						<?php echo HTMLHelper::_('date', $item->publish_up, Text::_('DATE_FORMAT_LC4')); ?>
+							<?php 
+								$date = $item->publish_up === $nulldate() ? $item->created : $item->publish_up;
```
Comment (Quy) [inline administrator/modules/mod_latest/tmpl/default.php]: Move after line 18?

Comment (chmst) [reply administrator/modules/mod_latest/tmpl/default.php]: NullDate is always the same, can be outside the loop

File: administrator/modules/mod_popular/tmpl/default.php (hunk 1 above)
Comment (Quy) [reply administrator/modules/mod_popular/tmpl/default.php]: I see it as assignments at the top and output of these variables in the body.

File: administrator/modules/mod_latest/tmpl/default.php (hunk 2 above)
Comment (Quy) [reply administrator/modules/mod_latest/tmpl/default.php]: I see it as assignments at the top and output of these variables in the body.

Comment (Quy) [reply administrator/modules/mod_latest/tmpl/default.php]: Nevermind.

File: administrator/modules/mod_latest/tmpl/default.php [hunk 3]
```diff
@@ -9,10 +9,13 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
 use Joomla\CMS\HTML\HTMLHelper;
 use Joomla\CMS\Language\Text;
 
 HTMLHelper::_('bootstrap.framework');
+
+$nulldate = Factory::getDbo()->getNullDate();
```
Comment (Quy) [inline administrator/modules/mod_latest/tmpl/default.php]: Please change to `$nullDate`.

**Comment by brianteeman** [comment]
Sorry but this looks to be a band aid solution and doesn't address the real issue as we are going to see the same problem throughout joomla and extensions wherever we use the publish_up date. Surely the fix is to return to the j3 method of setting the article publish_up date as equally to the created date by default instead of setting it to null.

**Comment by chmst** [comment]
What do you suggest to show in the modules if the publish_up date is null?

**Comment by brianteeman** [comment]
I just took a closer look and actually its complicated

The behaviour now in Joomla 4 is that if you create a new article with a status of unpublished then it is created with a null for publish_up
If you go in to the article and change the status to publish and then save and you get a publish_up of now
BUT If you change the status from the list view then you do not get a publish_up date at all

So I think there are two seperate bugs here.
1. The article doesnt always get a publish_up date when it is published
2. What to display for an article date in the module when it has no publish_up date

When Issue 1 is fixed then the only time you will not have a publish_up date is when it is not published. So in that scenario we should be displaying "Unpublished" - displaying two different dates (published/created) is not correct

**Comment by brianteeman** [comment]
OR it can always display created date

**Comment by richard67** [comment]
I would assume by the name of that module, Popular Articles, that it does not show unpublished articles at all, because how shall an article become popular if it has never been published, or how shall it become more popular if it ever had been published and then was unpublished. But that's maybe only my opinion.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24968">issues.joomla.org/tracker/joomla-cms/24968</a>.</sub>

**Comment by brianteeman** [comment]
You assume incorrectly. It might have been published in the past for example

**Comment by richard67** [comment]
Yes, and then be unpublished again. Question is if it then still makes sense to show it in the list of popular articles. I think that's a matter of taste.

Anyway, bug no. 1 with the status change in the list view to published does not set the publish date, this has to be fixed.

For 2. I think showing text "Unpublished" would be good for my taste, if that is possible.

**Comment by chmst** [comment]
Maybe showing the create date is the simplest and clearest solution. All other possible combinations need explanation. If the haeder says "Date created" instead of "Date" it is clear enough, no matter which history the article has gone through. 
Additionally, the items could get the icon for published/unpublished also on the 
modules.

**Comment by Quy** [comment]
Please fix conflicts and see related PRs regarding null value for dates.

**Comment by richard67** [comment]
@chmst If necessary I can advise regarding null date, just ping me if needed.

**Comment by richard67** [comment]
If an article is newly created with unpublished state, the publish_up time is null in PHP and NULL in SQL after recent merged PRs. Before it was set to (pseudo-)nulldates in that case. In this case the publish_up time should **NOT** be set to the created time. As soon as the article is published, publish_up should be set by the software, otherwise it is a bug and has to be fixed. An article newly created with published state should also have publish_up set. Unpublished articles should not be shown by these modules, as far as I understand. So I would recommend to test if the issue which shall be solved by this PR still is an issue on current 4.0-dev branch. If it still is, I will check and advise how to fix, otherwise if it is not an isue anymore, this issue can be closed.

**Comment by chmst** [comment]
Thank you for your time and information. The original issue (new itemwit date -0001 .. ) is solved in 4.0, 
So I'll close this ussue.

**Comment by richard67** [comment]
Sometimes we are lucky with J4 :smile:

**Pull request closed without merging**